### ISSUE TITLE
useState provides this behavior already

### ISSUE THREAD

**Issue by ehaynes99** [Andarist/use-constant#4]
https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily

`useState` accepts a default value OR a create function for the default value. If a function, it is evaluated only once. Thus, this entire pattern could simply be a convenience function:

```typescript
import * as React from 'react'

export default function useConstant<T>(fn: () => T): T {
  return React.useState<T>(fn)[0]
}
```

**Comment by Andarist** [comment]
Yes, that's true - I prefer doing this with ref here because I know that I won't ever need a second value of the useState, therefore I'm skipping allocating it (and also skip allocating temporary array which is returned from useState). In general ref approach is slightly cheaper, which doesn't matter much, but whatever.

You can compare what is done under the hood by [useRef](https://github.com/facebook/react/blob/35d2b3bb5e01d8c388cb17a68608184dc7300892/packages/react-dom/src/server/ReactPartialRendererHooks.js#L367-L381) and [useState](https://github.com/facebook/react/blob/35d2b3bb5e01d8c388cb17a68608184dc7300892/packages/react-dom/src/server/ReactPartialRendererHooks.js#L248-L336).

**Comment by dantman** [comment]
@ehaynes99 Sadly my reference has disappeared so I can't link to it. But I had a Twitter discussion on this topic with one of the React devs and they advised against using useState for constants. Because like useMemo, useState is also open to future potential optimization scenarios that could result in your assumptions about data being "constant" breaking.

I think the example optimization they gave was a scenario where React decides that memory usage from storing useState data affects performance more than the CPU cost of running an initializer function. Maybe in response to an OOM or memory limit? That is the scenario where they suggest the useMemo cache could be purged as well.

The idea was it's possible that in the future React could optimize a component whose state had never changed (i.e. the return value of the initializer should be the same thing as the current state of the component) by not storing state for unchanged states and instead just re-running the initializer when updating the component.

**Comment by ehaynes99** [comment]
I'm doubtful of that use case. It's impossible for React to determine that the initializer result will be the same, e.g.:
```javascript
const startTime = React.useState(() => Date.now())[0]
```
This would lead to behavioral changes, not just resource allocation. At any rate, they're not my assumptions; it's explicitly stated in the docs.

> useMemo lets you memoize an expensive calculation if the dependencies are the same. However, it only serves as a hint, and doesn’t guarantee the computation won’t re-run. But sometimes you need to be sure an object is only created once.
> ...
> To avoid re-creating the ignored initial state, we can pass a function to useState:
> ...
> React will only call this function during the first render. See the useState API reference.

**Comment by OliverJAsh** [comment]
React Compiler complains about the ref implementation of `useConstant`. This might be another reason to prefer the `useState` implementation.

 https://playground.react.dev/#N4Igzg9grgTgxgUxALhASwLYAcIwC4AEAVAQIZgEBKCpchAZjBBgQOQw12sA6Adn3gCeWBFQRgoAGzwAhCAA8APABUAfAQC8BYAQBuyAsoIBfPnwTyc+AgBME9UlIZRedNBF4EoYBAGEPYHikvHgqqgAU9LwG4QCUmurKsQZGwHwEBHABhBz0mmK0eAB03gjUeYpiEtJy8gSVRqrqcWaeBGh54QCEuUVwsBwh8WltGb39MIOEWjr6BFFxJulL-G0ceLCe4wMIIUW6fMYgxkA

> Error: Cannot access refs during render
> 
> React refs are values that are not needed for rendering. Refs should only be accessed outside of render, such as in event handlers or effects. Accessing a ref value (the `current` property) during render can cause your component not to update as expected (https://react.dev/reference/react/useRef)
> 
>    6 |   const ref = React.useRef < ResultBox < T >> ()
>    7 |
> >  8 |   if (!ref.current) {
>      |        ^^^^^^^^^^^ Cannot access ref value during render
>    9 |     ref.current = { v: fn() }
>   10 |   }
>   11 |

**Comment by josephsavona** [comment]
@OliverJAsh We've actually been meaning to add a validation against useState() where the setter is unused, since that is also either a no-op (if the initial value can't change) or is going to drop updates (if the initial value can change).
